### PR TITLE
openssl_legacy: init and use

### PR DESCRIPTION
**Pull request by ajs124**
###### Description of changes
Follow-up to #196906

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
I don't think we can catch this amount of rebuilds before branch-off, unless `@mweinelt` is OK with delaying it by a few days.  (`*-darwin` stdenvs get rebuilt)  Of course, that still doesn't prevent it from getting staged for both 21.11 and unstable.

**Comment by vcunat** [comment]
`python310Packages.ntlm-auth` builds with this, at least.

**Comment by mweinelt** [comment]
I think at this point we should go with staging-22.11.

**Comment by ajs124** [comment]
Alright, so I'll retarget this against staging and we backport it into staging-22.11, once that's created?

File: pkgs/development/libraries/openssl/default.nix [hunk 1]
```diff
@@ -192,6 +194,8 @@ let
       rm -r $etc/etc/ssl/misc
 
       rmdir $etc/etc/ssl/{certs,private}
+
+      ${lib.optionalString (conf != null) "cat ${conf} > $etc/etc/ssl/openssl.cnf"}
     '';
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/openssl/default.nix:199]:
```suggestion
    '' + lib.optionalString (conf != null) ''
      cat ${conf} > $etc/etc/ssl/openssl.cnf
    '';
```
Normally not a big fan of this but shouldn't this cut down the amount of rebuilds?

Comment (ajs124) [reply pkgs/development/libraries/openssl/default.nix:199]: Probably, but I'd guess not significantly, since this PR changes `openldap` and `python3`.

**Comment by vcunat** [comment]
This doesn't seem to work well on darwin, but I don't know why:
https://hydra.nixos.org/eval/1786377?filter=openldap#tabs-now-fail

**Comment by SuperSandro2000** [comment]
It doesn't like rc4

```
additional info: SASL(-1): generic failure: internal error: failed to init cipher 'rc4'
```

**Comment by vcunat** [comment]
Yes, I thought the need for RC4 was one of the reasons to try keeping older openssl in there.

**Comment by ajs124** [comment]
Hm. It works just fine on linux though. Maybe the openssl config file doesn't work the same way on darwin?

**Comment by vcunat** [comment]
We'll need some way forward.  This breaks really many builds (roughly 1000*2).

I could simply put the `rm` line back as `optionalString stdenv.isDarwin`, but perhaps it would be better if someone checked that we're not leaving some big problem in there.

**Comment by vcunat** [comment]
I'm not familiar with library loading on darwin.

But I believe that even even on Linux+glibc these library overrides are a bit difficult.  I'm sure `libldap.so` is typically not used stand-alone, and some binaries using it probably use openssl already (not in `_legacy` variant).  In those cases it's not clear which version is loaded first and thus wins.  What I'm certain of is that when you run an executable (without using extra loading magic), a particular SONAME is satisfied just once, including in all libraries &ndash; and thus each symbol name is consistent across the whole executable.

This can also cause problems when impure loading happens, e.g. in particular with `/run/opengl-driver`.

**Comment by ajs124** [comment]
Usage of openldap as an executable or library are both common

**Comment by vcunat** [comment]
This is in `nixos-unstable` now, so I assume we're still backporting to 22.11 even though it's long past release?

**Pull request merged**